File: ios/BsPatch.swift
import Foundation
import Compression

/// bspatch — applies a binary patch produced by the server's bsdiff.
///
/// Only the patch side lives on device; generating diffs is the server's job.
///
/// Patch layout (see packages/backend/src/utils/bsdiff.ts):
///
///     magic        8 bytes   "OTABSDF1"
///     ctrlLen      8 bytes   compressed length of the control block
///     diffLen      8 bytes   compressed length of the diff block
///     newSize      8 bytes   length of the reconstructed file
///     ctrl block   zlib      triples of (addLen, extraLen, oldSeek)
///     diff block   zlib      byte-wise deltas for the "add" runs
///     extra block  zlib      literal bytes with no match in old
///
/// Offsets are sign-magnitude, NOT two's complement: eight little-endian bytes
/// where bit 7 of the last byte carries the sign. Reading these as a regular
/// Int64 gives silently wrong values for the (frequently negative) oldSeek.
enum BsPatch {

    private static let magic = "OTABSDF1"
    private static let headerSize = 32

    enum PatchError: LocalizedError {
        case malformed(String)

        var errorDescription: String? {
            switch self {
            case .malformed(let reason): return "Invalid bsdiff patch: \(reason)"
            }
        }
    }

    /// Reconstruct the new file from `oldData` and `patchData`.
    static func apply(oldData: Data, patchData: Data) throws -> Data {
        guard patchData.count >= headerSize else {
            throw PatchError.malformed("patch is smaller than its header")
        }

        let header = [UInt8](patchData.prefix(headerSize))

        let foundMagic = String(bytes: header[0..<8], encoding: .ascii) ?? ""
        guard foundMagic == magic else {
            throw PatchError.malformed("bad magic: expected \(magic), got \(foundMagic)")
        }

        let ctrlLen = readOffset(header, 8)
        let diffLen = readOffset(header, 16)
        let newSize = readOffset(header, 24)

        guard ctrlLen >= 0, diffLen >= 0, newSize >= 0 else {
            throw PatchError.malformed("negative length in header")
        }
        guard newSize <= Int64(Int.max) else {
            throw PatchError.malformed("reconstructed size exceeds addressable range")
        }
        guard headerSize + Int(ctrlLen) + Int(diffLen) <= patchData.count else {
            throw PatchError.malformed("patch is truncated")
        }

        let ctrlStart = headerSize
        let diffStart = ctrlStart + Int(ctrlLen)
        let extraStart = diffStart + Int(diffLen)

        let ctrl = try inflate(patchData.subdata(in: ctrlStart..<diffStart), label: "control")
        let diff = try inflate(patchData.subdata(in: diffStart..<extraStart), label: "diff")
        let extra = try inflate(
            patchData.subdata(in: extraStart..<patchData.count), label: "extra"
        )

        let targetSize = Int(newSize)
        var newBytes = [UInt8](repeating: 0, count: targetSize)
        let oldBytes = [UInt8](oldData)
        let ctrlBytes = [UInt8](ctrl)
        let diffBytes = [UInt8](diff)
        let extraBytes = [UInt8](extra)

        var oldPos = 0
        var newPos = 0
        var ctrlPos = 0
        var diffPos = 0
        var extraPos = 0

        while newPos < targetSize {
            guard ctrlPos + 24 <= ctrlBytes.count else {
                throw PatchError.malformed("control block exhausted before output was full")
            }

            let addLen = Int(readOffset(ctrlBytes, ctrlPos))
            let extraLen = Int(readOffset(ctrlBytes, ctrlPos + 8))
            let oldSeek = Int(readOffset(ctrlBytes, ctrlPos + 16))
            ctrlPos += 24

            guard addLen >= 0, extraLen >= 0 else {
                throw PatchError.malformed("negative run length in control block")
            }
            guard newPos + addLen <= targetSize else {
                throw PatchError.malformed("add run overruns the output buffer")
            }
            guard diffPos + addLen <= diffBytes.count else {
                throw PatchError.malformed("diff block exhausted")
            }

            // Copy from old, adding the byte-wise delta. Positions outside the
            // old file contribute zero, matching the reference implementation.
            for i in 0..<addLen {
                let oldIndex = oldPos + i
                let oldByte: UInt8 = (oldIndex >= 0 && oldIndex < oldBytes.count)
                    ? oldBytes[oldIndex]
                    : 0
                newBytes[newPos + i] = diffBytes[diffPos + i] &+ oldByte
            }

            newPos += addLen
            oldPos += addLen
            diffPos += addLen

            guard newPos + extraLen <= targetSize else {
                throw PatchError.malformed("extra run overruns the output buffer")
            }
            guard extraPos + extraLen <= extraBytes.count else {
                throw PatchError.malformed("extra block exhausted")
            }

            if extraLen > 0 {
                newBytes.replaceSubrange(
                    newPos..<(newPos + extraLen),
                    with: extraBytes[extraPos..<(extraPos + extraLen)]
                )
            }
            newPos += extraLen
            extraPos += extraLen

            oldPos += oldSeek
        }

        return Data(newBytes)
    }

    /// Read a sign-magnitude 64-bit offset.
    private static func readOffset(_ buf: [UInt8], _ offset: Int) -> Int64 {
        var y = Int64(buf[offset + 7] & 0x7f)
        var i = 6
        while i >= 0 {
            y = y * 256 + Int64(buf[offset + i])
            i -= 1
        }
        return (buf[offset + 7] & 0x80) != 0 ? -y : y
    }

    /// Inflate a zlib stream.
    ///
    /// Apple's Compression framework speaks raw DEFLATE, not zlib, so the
    /// 2-byte zlib header and 4-byte Adler-32 trailer are stripped first.
    private static func inflate(_ data: Data, label: String) throws -> Data {
        guard data.count > 6 else {
            // Nothing but header and trailer — an empty block is legitimate
            // (e.g. no deletions), so return empty rather than failing.
            return Data()
        }

        // An empty block is common, not exceptional: a patch with no "extra"
        // bytes (every byte matched the base) compresses to an 8-byte stream
        // that inflates to nothing. compression_decode_buffer returns 0 for
        // both "produced no output" and "failed", so the empty case has to be
        // recognised up front or every such patch is rejected as corrupt.
        //
        // Adler-32 of an empty input is exactly 1, so a legitimately empty
        // stream always ends with 00 00 00 01.
        let trailer = data.suffix(4)
        if trailer.elementsEqual([0x00, 0x00, 0x00, 0x01]) {
            return Data()
        }

        let raw = data.subdata(in: 2..<(data.count - 4))

        // Output size is unknown up front. Grow the buffer until decompression
        // stops filling it completely.
        var capacity = max(raw.count * 4, 64 * 1024)

        for _ in 0..<8 {
            var output = Data(count: capacity)
            let written: Int = output.withUnsafeMutableBytes { outPtr -> Int in
                guard let outBase = outPtr.bindMemory(to: UInt8.self).baseAddress else {
                    return 0
                }
                return raw.withUnsafeBytes { inPtr -> Int in
                    guard let inBase = inPtr.bindMemory(to: UInt8.self).baseAddress else {
                        return 0
                    }
                    return compression_decode_buffer(
                        outBase, capacity,
                        inBase, raw.count,
                        nil, COMPRESSION_ZLIB
                    )
                }
            }

            if written == 0 {
                throw PatchError.malformed("\(label) block failed to decompress")
            }
            if written < capacity {
                return output.prefix(written)
            }
            // Filled the buffer exactly — output may have been truncated.
            capacity *= 4
        }

        throw PatchError.malformed("\(label) block is implausibly large")
    }
}

File: android/src/main/java/com/otasdk/BsPatch.kt
package com.otasdk

import java.io.ByteArrayOutputStream
import java.util.zip.Inflater

/**
 * bspatch — applies a binary patch produced by the server's bsdiff.
 *
 * Only the patch side lives on device; generating diffs is the server's job.
 *
 * Patch layout (see packages/backend/src/utils/bsdiff.ts):
 *
 *   magic        8 bytes   "OTABSDF1"
 *   ctrlLen      8 bytes   compressed length of the control block
 *   diffLen      8 bytes   compressed length of the diff block
 *   newSize      8 bytes   length of the reconstructed file
 *   ctrl block   zlib      triples of (addLen, extraLen, oldSeek)
 *   diff block   zlib      byte-wise deltas for the "add" runs
 *   extra block  zlib      literal bytes with no match in old
 *
 * Offsets are sign-magnitude, NOT two's complement: eight little-endian bytes
 * where bit 7 of the last byte carries the sign. Reading these as a normal
 * signed long produces silently wrong values for the (frequently negative)
 * oldSeek field.
 */
object BsPatch {

    private const val MAGIC = "OTABSDF1"
    private const val HEADER_SIZE = 32

    class PatchFormatException(message: String) : Exception(message)

    /**
     * Reconstruct the new file from [oldData] and [patchData].
     *
     * @throws PatchFormatException if the patch is malformed or truncated.
     */
    @JvmStatic
    fun apply(oldData: ByteArray, patchData: ByteArray): ByteArray {
        if (patchData.size < HEADER_SIZE) {
            throw PatchFormatException("Patch is smaller than its header")
        }

        val magic = String(patchData, 0, 8, Charsets.US_ASCII)
        if (magic != MAGIC) {
            throw PatchFormatException("Bad magic: expected $MAGIC, got $magic")
        }

        val ctrlLen = readOffset(patchData, 8)
        val diffLen = readOffset(patchData, 16)
        val newSize = readOffset(patchData, 24)

        if (ctrlLen < 0 || diffLen < 0 || newSize < 0) {
            throw PatchFormatException("Negative length in patch header")
        }
        if (newSize > Int.MAX_VALUE) {
            throw PatchFormatException("Reconstructed size $newSize exceeds addressable range")
        }
        if (HEADER_SIZE + ctrlLen + diffLen > patchData.size) {
            throw PatchFormatException("Patch is truncated")
        }

        val ctrlStart = HEADER_SIZE
        val diffStart = (ctrlStart + ctrlLen).toInt()
        val extraStart = (diffStart + diffLen).toInt()

        val ctrl = inflate(patchData, ctrlStart, ctrlLen.toInt(), "control")
        val diff = inflate(patchData, diffStart, diffLen.toInt(), "diff")
        val extra = inflate(patchData, extraStart, patchData.size - extraStart, "extra")

        val newData = ByteArray(newSize.toInt())
        var oldPos = 0L
        var newPos = 0
        var ctrlPos = 0
        var diffPos = 0
        var extraPos = 0

        while (newPos < newSize) {
            if (ctrlPos + 24 > ctrl.size) {
                throw PatchFormatException("Control block exhausted before output was full")
            }

            val addLen = readOffset(ctrl, ctrlPos)
            val extraLen = readOffset(ctrl, ctrlPos + 8)
            val oldSeek = readOffset(ctrl, ctrlPos + 16)
            ctrlPos += 24

            if (addLen < 0 || extraLen < 0) {
                throw PatchFormatException("Negative run length in control block")
            }
            if (newPos + addLen > newSize) {
                throw PatchFormatException("Add run overruns the output buffer")
            }
            if (diffPos + addLen > diff.size) {
                throw PatchFormatException("Diff block exhausted")
            }

            // Copy from old, adding the byte-wise delta. Positions outside the
            // old file contribute zero, matching the reference implementation.
            for (i in 0 until addLen.toInt()) {
                val oldIndex = oldPos + i
                val oldByte = if (oldIndex in 0 until oldData.size.toLong()) {
                    oldData[oldIndex.toInt()].toInt() and 0xff
                } else {
                    0
                }
                val delta = diff[diffPos + i].toInt() and 0xff
                newData[newPos + i] = ((delta + oldByte) and 0xff).toByte()
            }

            newPos += addLen.toInt()
            oldPos += addLen
            diffPos += addLen.toInt()

            if (newPos + extraLen > newSize) {
                throw PatchFormatException("Extra run overruns the output buffer")
            }
            if (extraPos + extraLen > extra.size) {
                throw PatchFormatException("Extra block exhausted")
            }

            System.arraycopy(extra, extraPos, newData, newPos, extraLen.toInt())
            newPos += extraLen.toInt()
            extraPos += extraLen.toInt()

            oldPos += oldSeek
        }

        return newData
    }

    /** Read a sign-magnitude 64-bit offset. */
    private fun readOffset(buf: ByteArray, offset: Int): Long {
        var y = (buf[offset + 7].toLong() and 0x7f)
        for (i in 6 downTo 0) {
            y = y * 256 + (buf[offset + i].toLong() and 0xff)
        }
        return if ((buf[offset + 7].toInt() and 0x80) != 0) -y else y
    }

    private fun inflate(src: ByteArray, offset: Int, length: Int, label: String): ByteArray {
        if (length < 0 || offset + length > src.size) {
            throw PatchFormatException("$label block extends past the end of the patch")
        }

        val inflater = Inflater()
        try {
            inflater.setInput(src, offset, length)
            val out = ByteArrayOutputStream(maxOf(length * 4, 1024))
            val buffer = ByteArray(64 * 1024)

            while (!inflater.finished()) {
                val n = inflater.inflate(buffer)
                if (n == 0) {
                    if (inflater.needsInput() || inflater.needsDictionary()) break
                }
                out.write(buffer, 0, n)
            }
            return out.toByteArray()
        } catch (e: Exception) {
            throw PatchFormatException("$label block failed to decompress: ${e.message}")
        } finally {
            inflater.end()
        }
    }
}
